lib/stream/cache.go

```diff
@@ -0,0 +1,172 @@
+package stream
+
+import (
+	"sync"
+)
+
+func NewNopCache[E any]() Cache[E] {
+	return new(nopCache[E])
+}
+
+type nopCache[E any] struct{}
+
+func (nopCache[E]) Get(uint64) []E  { return nil }
+func (nopCache[E]) Set(uint64, []E) {}
+
+//
+// func NewSliceCache[E any](limit int) Cache[E] {
```

Comment:
I implemented a slice based cache, but went for map in the end since it is much simpler and performance is acceptable.

Reply:
I'll remove this before finally merging the PR, it is just here for reference